### ISSUE TITLE
Generating derived graph from existing graph

### ISSUE THREAD

**Issue by msakuta** [msakuta/rustograd#7]
One way to achieve higher order differentiation is to use higher order dual numbers, which was demonstrated in PR #5, but it has limitation that it only works in forward mode.

# Deriving a graph from existing one

It should be possible to construct a new (sub) graph from existing values by applying transformations to the nodes.
It seems possible with existing AD libraries, such as [PyTorch](https://qiita.com/tmasada/items/9dee38e5bc1482217493) and [Jax](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#jvps-in-math).

A simple example is like this:

```rust
    let tape = Tape::new();
    let a = tape.term("a", 123.);
    let b = tape.term("b", 321.);
    let aba = a + b + a;
```

First, it constructs an expression `(a + b) + a`.
Then we can apply a transformation by `gen_graph` method:

```rust
    let daba = aba.gen_graph(&a).unwrap();
```

Now we have 2 graphs (that happens to be disjoint, but in general they should be able to share a subgraph):

![image](https://github.com/msakuta/rustograd/assets/2798715/1ad4dddc-6447-4017-a654-19e934c01b2d)

Note that the _data_ of `d a + d a` is 2, which is the same as the _grad_ field of `a` (which is derivative of `((a + b) + a)` with respect to `a`).

Also note that further applying `gen_graph` will yield nothing, because the expression is a linear function to `a`, more than 1st order differentiation will be 0.

```rust
assert!(daba.gen_graph(&a).is_none());
```

Things get more interesting when we have a product.
This time, the expression is `(a - b) * a`.

```rust
    let tape = Tape::new();
    let a = tape.term("a", 123.);
    let b = tape.term("b", 321.);
    let aba = (a - b) * a;
    let daba = aba.gen_graph(&a).unwrap();
    aba.eval();
    aba.backprop().unwrap();
    daba.eval_noclear();
```

The product rule brings partially shared expression graph.

![image](https://github.com/msakuta/rustograd/assets/2798715/cacb0ea9-5da0-48c6-9de4-adb5ccff2a4f)

It also allows higher order differentiation when applied multiple times.
For example, you can get arbitrary order differentiation of Gaussian distribution like below.

![many-derive](https://github.com/msakuta/rustograd/assets/2798715/e5b4b0c7-a6ee-4858-a514-3eeee991043f)

Note that the computational graph can get complicated pretty quickly, because differentiating a Gaussian produces multiple terms, and they will grow up cascadingly. It is probably not the most efficient way of calculating higher order differentiation.

![image](https://github.com/msakuta/rustograd/assets/2798715/166d1d1c-8069-4341-a097-2453171d55df)

We can visualize how the graph grows by an animation, but it's not that helpful to understand what's going on, because the node positions change everytime by graphviz.
However, you may notice that many duplicate nodes are created, such as -1 and -a. This is because we do not de-duplicate in the processof generating subgraph, because it could be costly to scan all existing nodes to find duplicates.

![gen_graph](https://github.com/msakuta/rustograd/assets/2798715/cc02ae59-ff88-4270-8aaa-9e4e9f5bbc5c)

Another example: sine wave.

![many-derive](https://github.com/msakuta/rustograd/assets/2798715/65890ef1-4b38-48ac-ae9a-6f88afd09360)

It happens to be simple compute graph, since sine wave does not produce polynomials

![image](https://github.com/msakuta/rustograd/assets/2798715/6f2a97b6-b4a9-484c-bb7c-e2bdf23a5511)

Yet another example is a sigmoid function.

![sigmoid-derive](https://github.com/msakuta/rustograd/assets/2798715/fa8b70f4-5205-4201-af8e-91fed00d0d88)

![image](https://github.com/msakuta/rustograd/assets/2798715/81f76eb2-1541-432a-86d5-14b2a7ce6128)

# How to achieve this semantics?

We can implement a graph generation algorithm out from existing ones.
To do that, we define a function `gen_graph` that is similar to `derive`.

```rust
fn gen_graph<T: Tensor>(nodes: &mut Vec<TapeNode<T>>, idx: u32, wrt: u32) -> Option<u32> {
    use TapeValue::*;
    match nodes[idx as usize].value {
        _ => todo!(),
    }
}
```

Inside the match clause, we can write transformation rules of each node.
The most basic node type is the `Value`. If it was the target variable to derive, we generate a constant node with a value 1.
If not, we do not generate a node.

```rust
        Value(_) => {
            if idx == wrt {
                let new_name = format!("d {}", nodes[idx as usize].name);
                Some(add_node(nodes, new_name, Value(T::one())))
            } else {
                None
            }
        }
```

The next type is `Add` or `Sub`.
If either one of the inputs yields a node, we use that node as the value to return.
If both of the inputs has some dependency on the variable `wrt`, we build a new node to add them.

```rust
        Add(lhs, rhs) => {
            let lhs = gen_graph(nodes, derive_map, lhs, wrt);
            let rhs = gen_graph(nodes, derive_map, rhs, wrt);
            match (lhs, rhs) {
                (Some(lhs), None) => Some(lhs),
                (None, Some(rhs)) => Some(rhs),
                (Some(lhs), Some(rhs)) => Some(add_add(nodes, lhs, rhs)),
                _ => None,
            }
        }
```

Subtraction is almost the same, but the right hand side will be negated.

```rust
        Sub(lhs, rhs) => {
            let lhs = gen_graph(nodes, derive_map, lhs, wrt);
            let rhs = gen_graph(nodes, derive_map, rhs, wrt);
            match (lhs, rhs) {
                (Some(lhs), None) => Some(lhs),
                (None, Some(rhs)) => Some(add_neg(nodes, rhs)),
                (Some(lhs), Some(rhs)) => Some(add_sub(nodes, lhs, rhs)),
                _ => None,
            }
        }
```

The `Mul` operation is a bit more complicated, because it needs 4 nodes to join at the maximum.
If one of the terms are dependent on a node, we can use the other value as the new node's `Mul` factor.
It is a direct translation of the product rule of differentiation.

```rust
        Mul(lhs, rhs) => {
            let dlhs = gen_graph(nodes, derive_map, lhs, wrt);
            let drhs = gen_graph(nodes, derive_map, rhs, wrt);
            match (dlhs, drhs) {
                (Some(dlhs), None) => Some(add_mul(nodes, dlhs, rhs)),
                (None, Some(drhs)) => Some(add_mul(nodes, lhs, drhs)),
                (Some(dlhs), Some(drhs)) => {
                    let plhs = add_mul(nodes, dlhs, rhs);
                    let prhs = add_mul(nodes, lhs, drhs);
                    let node = add_add(nodes, plhs, prhs);
                    Some(node)
                }
                _ => None,
            }
        }
```

Division:

```rust
        Div(lhs, rhs) => {
            let dlhs = gen_graph(nodes, derive_map, lhs, wrt);
            let drhs = gen_graph(nodes, derive_map, rhs, wrt);
            match (dlhs, drhs) {
                (Some(dlhs), None) => Some(add_div(nodes, dlhs, rhs)),
                (None, Some(drhs)) => {
                    let node = add_mul(nodes, lhs, drhs);
                    let node = add_div(nodes, node, rhs);
                    let node = add_div(nodes, node, rhs);
                    Some(add_neg(nodes, node))
                }
                (Some(dlhs), Some(drhs)) => {
                    let plhs = add_div(nodes, dlhs, rhs);
                    let node = add_mul(nodes, lhs, drhs);
                    let prhs = add_div(nodes, node, rhs);
                    let prhs = add_div(nodes, prhs, rhs);
                    Some(add_sub(nodes, plhs, prhs))
                }
                _ => None,
            }
        }
```

Negation:

```rust
Neg(term) => gen_graph(nodes, derive_map, term, wrt).map(|node| add_neg(nodes, node)),
```

UnaryFn:

Now, this is complicated, because an arbitrary unary function can produce a node on either input, output or derived node of its parent node. We allow the implementor of `UnaryFn` trait to customize the behavior by calling `gen_graph` trait method.

```rust
        UnaryFn(UnaryFnPayload { term, ref mut f }) => {
            let taken_f = f.take();
            let derived = gen_graph(nodes, derive_map, term, wrt);
            let ret = derived.and_then(|derived| {
                taken_f
                    .as_ref()
                    .unwrap()
                    .gen_graph(nodes, idx, term, derived)
            });
            if let UnaryFn(UnaryFnPayload { ref mut f, .. }) = nodes[idx as usize].value {
                *f = taken_f;
            } else {
                unreachable!()
            }
            ret
        }
```

## Implementing UnaryFn

Implementing UnaryFn can be done by the user. For example, sine wave can be defined as below.

```rust
struct SinFn(usize);

impl UnaryFn<f64> for SinFn {
    fn name(&self) -> String {
        match self.0 % 4 {
            0 => "sin",
            1 => "cos",
            2 => "-sin",
            3 => "-cos",
            _ => unreachable!(),
        }
        .to_string()
    }

    fn f(&self, data: f64) -> f64 {
        match self.0 % 4 {
            0 => data.sin(),
            1 => data.cos(),
            2 => -data.sin(),
            3 => -data.cos(),
            _ => unreachable!(),
        }
    }

    fn grad(&self, data: f64) -> f64 {
        Self(self.0 + 1).f(data)
    }

    fn gen_graph(
        &self,
        nodes: &mut Vec<TapeNode<f64>>,
        _idx: u32,
        input: u32,
        _derived: u32,
    ) -> Option<u32> {
        Some(add_unary_fn(nodes, Box::new(Self(self.0 + 1)), input))
    }
}
```

The key point here is that the `SinFn` struct can have a payload, so it can customize the behavior depending on the order of differentiation.
Theoretically, `-sin` and `-cos` are not necessary, since you can combine them with negation operation, but it would increase the number of nodes and would be less efficient.
